test/common/upstream/BUILD

```diff
@@ -127,6 +127,7 @@ envoy_cc_benchmark_binary(
 
 envoy_benchmark_test(
     name = "eds_speed_test_benchmark_test",
+    size = "large",
```

Comment:
Prefer factoring out to distinct test/benchmark binary to save the CO2.

Reply:
> Prefer factoring out to distinct test/benchmark binary to save the CO2.

The issue is the > 10x time increase under tsan.  Breaking this into 2 tests would still consume the same number of computrons.

Reply:
Counteroffer: only run 100 iterations or so (and no rang) in the checked in version, and leave a comment that 100,000 is a reasonable place to start if you're trying to measure something?

Reply:
I did this and the fastbuild runtime is <2s

Reply:
Is it still possible to observe the difference in performance with/without your other changes when the number of endpoints is set so small?

Can we set this to something in between?

Reply:
Yeah I think we need to have two different modes. What's the problem with doing that?

Reply:
There is a way to provide a custom function that returns the list of thresholds to use for the benchmark.  Using that combined with a flag to reduce size when just doing the smoke test would be helpful.  That said, the duality of benchmark binary and test has caused some confusion so far; people tend to just run the test.

Reply:
having a single binary would be great, as long as the default mode runs fast and it's super-easy for users to understand how to make the perf-test do something interesting (which will generally be slower than we'd want in CI).

Reply:
See Benchmark::Apply

Use that instead of Benchmark::Arg when instantiating the benchmark.

Reply:
repeating back the proposal to be sure I understand:
/upstream:eds_speed_test takes a flag, when true it runs a 100,000 nodes, when false some trivial number. /upstream:eds_speed_test_benchmark_test sets that flag to false.

The case I am (personally) trying to optimize for is not slowing down someone who runs baze test //test/...:all, and not expending compute resources unnecessarily

Reply:
yes that sounds good to me. I haven't done this before myself; as I never tried to make benchmarks to run in CI but this would seem like the right way to do it, and I had been given the impression it worked that way for the header-map speed tests. But that pattern sounds good.

Reply:
pgenera: yeah, that's the idea.  Have the flag be default to the more expensive benchmark, but disable it when running the benchmark smoke test.

Reply:
Okay, having stared at this for a while it seems that the [Supported Way](https://github.com/google/benchmark/blob/c078337494086f9372a46b4ed31a3ae7b3f1a6a2/include/benchmark/benchmark.h#L42) to pass arbitrary command line flags to a google/benchmark test is to copy the main() from the [BENCHMARK_MAIN macro](https://github.com/google/benchmark/blob/c078337494086f9372a46b4ed31a3ae7b3f1a6a2/include/benchmark/benchmark.h#L1276) into your test and parse out the flag before calling Initialize(argc, argv). This is because the default main returns an error if it encounters any unknown flags.

I'm going to do this, but it seems weird enough that I thought I'd say it out loud before diving in.

Reply:
Quick status update:

- `TCLAP` won't work because it wants the flags to be defined in the same place as the parser object.
- `absl::flags` might work since the parser finds defined flags via reflection, but it won't build with -Wold_style_cast and I'm not sure there's a way to programatically override `--undefok` so that we can still use benchmark's flag parsing for its flags.

So I'm at a bit of a dead end with what felt like an easy request; I'm currently investigating why `//test/common/http:header_map_impl_speed_test` behaves differently than eds_speed_test despite using the same macros and build target types (the former returns immediately in bazel test, the latter runs all its tests).

If I've completely lost the thread, or I'm chasing diminishing returns, I'd be happy to get this in to capture the performance improvements and continue working on eds_speed_test improvements in a followup.

Reply:
Hey @antoniovicente; I think this now reflects what we discussed; however:

- I tried and failed to do the modification in an Apply() function to the benchmarks; runtime changes to the variables were ignored as if the functions were only being executed when the macro expansion happened.
- Is this use of a global okay? It doesn't feel okay, but I failed to come up with anything better.

In any case, eds_speed_test_benchmark_test (fastbuild) now completes in <3s locally, while running eds_speed_test directly takes 40s (opt). I think this reflects how users and CI will interact with the tests.